pkg/operator/targetconfigcontroller/targetconfigcontroller.go

```diff
@@ -94,6 +94,19 @@ func (c TargetConfigController) sync(ctx context.Context, syncCtx factory.SyncCo
 	if len(envVars) == 0 {
 		return fmt.Errorf("TargetConfigController missing env var values")
 	}
+        checkEnvNum := 0
+        for _, k := range sets.StringKeySet(envVars).List() {
+            if strings.Contains(k, "_ETCD_URL_HOST"){
+                checkEnvNum++
+            }else if strings.Contains(k, "_ETCD_NAME"){
+                checkEnvNum++
+            }else if strings.Contains(k, "_IP"){
+                checkEnvNum++
+            }
+        }
+        if checkEnvNum < 3 {
+            return fmt.Errorf("TargetConfigController mismatch checkEnvNum")
```

Comment:
besides the formatting issues, I don't really understand what the purpose of this is.
The variables are generated in etcd_env.go directly from the node lister and the static pod machinery:

https://github.com/openshift/cluster-etcd-operator/blob/b561d08d887fc0916ba6008308b8621754e30302/pkg/etcdenvvar/etcd_env.go#L172-L193


if there are none of those we need to fix this up the stack in https://github.com/openshift/library-go and not here. 

Do you know why your static pod operator status does not contain the required node status?

Reply:
not to say that I'm opposed against validation, but maybe we should add this in the place where it occurs, for example here:

https://github.com/openshift/cluster-etcd-operator/blob/b561d08d887fc0916ba6008308b8621754e30302/pkg/etcdenvvar/etcd_env.go#L142

so if the length of `envVarContext.status.NodeStatuses` is zero, this should error out directly.

Reply:
yes ,Under normal circumstances, the EtcdURLHost is generated by the EnvVarController, and the TargetConfigController will use the GetEnvVars to get the env to generate etcd yaml

but in some cases，the GetEnvVars is not zero, bu not containes "_ETCD_URL_HOST", "_ETCD_NAME","_IP"; but TargetConfigController  will still generate etcd yaml, ans this yaml will checked fail by etcd-ensure-env-vars init container like ""#!/bin/sh\nset -euo pipefail\n\n: \"${NODE_master0_ETCD_URL_HOST?not set}\"\n: \"${NODE_master0_ETCD_NAME?not set}\"\n: \"${NODE_master0_IP?not set}\"

Reply:
like this:
<img width="589" alt="image" src="https://user-images.githubusercontent.com/30207518/217510999-113d1d41-8335-4c33-8c9b-e2b363818eb6.png">

Reply:
okay, I think there's a bigger bug we need to go after in the static pod machinery. Which OpenShift version are you running? 

Could also you please add your validation directly into the etcd_env.go on the respective functions? I think that's the better place, all other controllers will benefit from this too.

Reply:
i am using ocp 4.9 and 4.11, both have the problem

Reply:
I have read all the ceo code, and GetEnvVars is only used by the TargetConfigController, and my modification will resolve this problem

Reply:
> 

I have read the ceo code, and the GetEnvVars is only used by the TargetConfigController; as u said"add your validation directly into the etcd_env.go", i think the problem is not in the EnvVarController, but in the timing when should TargetConfigController use the env, TargetConfigController  should wait all the env it needs to be ready, and then it can generate the etcd yaml

Reply:
I have read the code too, yet I would like to avoid a case where other people add a new controller that requires env vars and read incomplete data. 

> TargetConfigController should wait all the env it needs to be ready, and then it can generate the etcd yaml

If this is a race condition between the controllers we need to add some locking or serialization for them, not add some if conditions on the consumer side of the API. I'll see if I can reproduce this and will come back to you on this.

Reply:
I think we can also add check in following:
https://github.com/openshift/cluster-etcd-operator/blob/f695ca59f559e6b7a23451f42e1d430aa1d4e380/pkg/etcdenvvar/etcd_env.go#L156

we can check the nodeInfo should exist, will also resolve this

Reply:
like this env get :
https://github.com/openshift/cluster-etcd-operator/blob/f695ca59f559e6b7a23451f42e1d430aa1d4e380/pkg/etcdenvvar/etcd_env.go#L204

Reply:
I have changed the etcdenv check func, please review the new code

Reply:
sounds good, I just added some more tests around this in https://github.com/openshift/cluster-etcd-operator/pull/1002

can you please take a look at that one?

Reply:
ok